### ISSUE TITLE
Add support for JWKS

### ISSUE THREAD

**Issue by toloco** [open-policy-agent/opa#1211]
# Expected Behavior
To be able to decode JWT with oi.jwt.decode_verify via JWKS certificate

# Actual Behavior
Currently io.jwt.decode_verify only supports secret and certificate method

# Additional Info
Ideally, we would use a bundle to download latest certificate from a "### .well-known/openid-configuration" to download jwks certificate list, and the using that data be able to decode and verify a given JWT.

**Comment by alindeman** [comment]
I started scoping this out a bit, and since I'm relatively new to OPA development, I'd like some validation on my thoughts:

First, I think we need some way to get valid keys into `data`. My understanding is the options are similar to options 2-4 from [Best Practices: Identity and User Attributes](https://www.openpolicyagent.org/docs/best-practice-identity.html). Basically, the JWKS keys could be injected via bundles, push or pull during eval. Is that right? What I'm not as clear on is how bundles would work in this scenario: would there need to be an intermediate service purpose built to translate a JWKS endpoint into a bundle tarball that OPA could ingest? Is there any prior art for a service like this that I could use as a guide?

Next, it looks to me like `io.jwt.decode_verify` built-in can only accept PEM-formatted certificates, whereas JWKS are will return raw RSA or ECDSA parameters like `n`/`e` or `x`/`y`/`crv`. If we are using an intermediate service, that service could translate into PEM formatted keys. But if the JWKS keys are pulled directly (`http.send`), there'd need to be some function in OPA itself do that. Or, the built-in could be extended to support raw parameters. Does anyone have an early opinion on this?

I think given a list of keys pulled or pushed into `data`, it should be possible to string together something using built-ins that verifies a JWT signed a key from the list identified in the JOSE header. However, it might be worth introducing new built-ins to make the pattern easier in the common case. I can probably propose something once I have a better handle on the first two questions.

**Comment by repenno** [comment]
I have been working on this for some time so we should talk if you are interested

**Comment by alindeman** [comment]
> I have been working on this for some time so we should talk if you are interested

@repenno Sure, how can I help?

**Comment by tsandall** [comment]
Extending `io.jwt.decode_verify` to support alternative formats like JWKS sounds reasonable to me. This is a (relatively) conservative addition that composes nicely with options 2-4 from that document. In the future we could explore more sophisticated built-ins that fetch the JWK sets on the fly. The same goes for processing JOSE headers. Thoughts?

**Comment by repenno** [comment]
@alindeman I will commit what I have to a branch and we can discuss. 

I spent a considerable amount of time evaluating GoLang JWT libraries that can read, sign and encrypt JWK/JWS. I looked at Square's JOSE,  JWT-Go (https://github.com/dgrijalva/jwt-go) and JWX-Go (https://github.com/lestrrat-go/jwx).  I chose the latter to add to OPA's vendor directory. It is small, clean, very much customizable and has a MIT license.

I decided to support at minimum the algorithms and JWK from https://tools.ietf.org/html/rfc7515 and https://tools.ietf.org/html/rfc7519. We can go crazy later if needed.

Today I have JWX-Go integrated and a series of unit tests covering all the examples in RFC7515 and RFC7519. This includes parsing, signing and verifying JWTs  both based on JWK and PEM (from, say, a file or env variable).

So, what is missing you ask? I decided to add all supporting code to get JWK, JWS and JWE but _not_ the command extension themselves and their processing. More unit tests are needed for JWE and we need to think how many different ways we will accept JWKs and protected headers: JSON (certainly), bytes (?), files (unit tests already cover that), environment variables (that what we need to TLS), etc. We also need to add lots of manual examples.

thanks,

**Comment by tsandall** [comment]
Since OPA is embedded as a library in many use cases, we need to avoid thirdparty dependencies. Transitive dependencies introduce potential for version conflicts which we don't want to deal with.

In this case, the cost of adding a thirdparty dependency is not worth it.

**Comment by lstoll** [comment]
@tsandall is the suggestion then to greenfield implement/fork an implementation of the various JWT* in to this codebase? That seems problematic in it’s own way too. The codebase already has several dependencies, I what is the deciding criteria for any particular dependency?

Has any thought been put in to ways to potentially handle optional/opt-in functionality?

**Comment by repenno** [comment]
@lstoll That's pretty much what I told @tsandall offline. We have three feature requests for JWT: #1174 #1131 and this one. This amounts to pretty writing an entire JWT suite from scratch (it has lots of problems on its own), levering a library or forking and adapting a library. 

The library that I proposed is small, clean and easy to integrate. The main person is also very active and responds quickly. I fixed two issue during my evaluation: https://github.com/lestrrat-go/jwx
